### PR TITLE
Feature: NSSL Microphysics support

### PR DESCRIPTION
**Pull request by MicroTed**
## Description

Supports CCPP with new NSSL microphysics scheme, which has separate hydrometeor categories for graupel and hail (Mansell et al. 2010, JAS). This PR allows fv3atm to compile correctly with CCPP+NSSL, but a further update to atmos_cubed_sphere is needed to run correctly. The NSSL scheme is used in NSSL's daily FV3 convection-allowing forecasts, and the scheme itself is aimed at this scale (dx <= 4km).  Testing is underway also with a preliminary Warn-on-Forecast FV3 ensemble.

The NSSL scheme has namelist options to enable/disable the hail category (i.e., can be 6- or 7-class) and to enable/disable prediction of idealized CCN number concentration.

## Testing

Tested on unix (NOAA-Jet/Hera; Intel compilers) and MacOS (gcc-10 with hpc-stack)

Have the ufs-weather-model regression test been run?
No, but attempted on jet.

## Dependencies

Related (dependent) PR for CCPP physics (adding the NSSL microphysics): 
  https://github.com/NCAR/ccpp-physics/pull/761
Related (dependent for correct operation) PR for atmos_cubed_sphere (dev/emc branch): 
  https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/171
Related (nondependent) PR for Single Column Model: 
 https://github.com/NCAR/ccpp-scm/pull/277

### CODE REVIEW

**Comment by grantfirl** [comment]
@MicroTed Are you planning on submitting a PR for ufs-weather-model? If not, I can create one and potentially add one or more UFS regression tests that exercise NSSL MP using one the new suites that you've provided here.

I can also try to run existing RTs on Hera if you'd like, just let me know.

**Comment by MicroTed** [comment]
I will need to work on a PR for the regional workflow. I guess I'm not sure what would need to updated at the ufs-weather-model level. Would that be to add a suite to the RT? If you don't mind helping, that would be awesome. I tried to run an RT on jet yesterday, but I haven't quite figured it out (some runtime error on the first test).

**Comment by grantfirl** [comment]
> I will need to work on a PR for the regional workflow. I guess I'm not sure what would need to updated at the ufs-weather-model level. Would that be to add a suite to the RT? If you don't mind helping, that would be awesome. I tried to run an RT on jet yesterday, but I haven't quite figured it out (some runtime error on the first test).

We'll need a ufs-weather-model PR for at least updating the fv3atm submodule and for storing UFS RT logs. Adding a RT that uses NSSL is another reason, which I would highly suggest to do so that nobody breaks NSSL MP functionality accidentally in the future. 

I can do this and try to run RTs on Hera.

**Comment by grantfirl** [comment]
@MicroTed One other thing that needs to go into this pull request is to edit the .gitmodules file to point to your fork and PR branch of ccpp-physics for testing (see, e.g. https://github.com/NOAA-EMC/fv3atm/pull/471/files). You'll also need to point to your PR branch of gfdl_atmos_cubed_sphere in that file. UFS code managers want to be able to recursively checkout the top-level repository and get all of the associated code changes for testing.

**Comment by MicroTed** [comment]
Ah, OK, thanks for the tip! I'm not quite an expert on all things git yet... I'll look at those .gitmodule files.

As for the ufs-weather-model/tests, I see it will need appropriate inputs in the 'parm' directory (field_table, etc.). I can look at that, too.

**Comment by grantfirl** [comment]
OK, I've opened a draft PR for ufs-weather-model: https://github.com/ufs-community/ufs-weather-model/pull/1032
It can go live when you're ready.

Yep, we may need a new namelist in ufs-weather-model/tests/parm. I am thinking that perhaps the new NSSL RT can be based on the existing rrfs_v1beta test. If we need a new field_table, then we'll need to ask UFS code managers to stage that on the test platforms. Other than the field_table, does NSSL need any special input data to run that we also need to stage?

**Comment by grantfirl** [comment]
@MicroTed I've submitted PRs into your PR branches of ccpp-physics, fv3atm, and ccpp-scm to fix some minor things caught during the UFS RTs. Please merge at your convenience if you are OK with the fixes.

With those fixes, I ran the full UFS regression tests on Hera/Intel. All tests ran. Most tests were bit-for-bit. However, there was a set of tests the failed in "check_result" for the regression tests, which mean that they changed the answer somehow. We need to understand why the answer was changed in these tests in order to proceed. The list of failed tests is:
regional_control 036 failed in check_result
regional_control 036 failed in run_test
regional_noquilt 038 failed in check_result
regional_noquilt 038 failed in run_test
regional_2threads 039 failed in check_result
regional_2threads 039 failed in run_test
regional_hafs 040 failed in check_result
regional_hafs 040 failed in run_test
regional_netcdf_parallel 041 failed in check_result
regional_netcdf_parallel 041 failed in run_test
regional_RRTMGP 042 failed in check_result
regional_RRTMGP 042 failed in run_test
rap_control 043 failed in check_result
rap_control 043 failed in run_test
rap_2threads 044 failed in check_result
rap_2threads 044 failed in run_test
rap_sfcdiff 046 failed in check_result
rap_sfcdiff 046 failed in run_test
hrrr_control 048 failed in check_result
hrrr_control 048 failed in run_test
rrfs_v1beta 049 failed in check_result
rrfs_v1beta 049 failed in run_test
control_csawmg 054 failed in check_result
control_csawmg 054 failed in run_test
control_csawmgt 055 failed in check_result
control_csawmgt 055 failed in run_test
control_thompson 058 failed in check_result
control_thompson 058 failed in run_test
control_thompson_no_aero 059 failed in check_result
control_thompson_no_aero 059 failed in run_test
control_csawmg_debug 068 failed in check_result
control_csawmg_debug 068 failed in run_test
control_csawmgt_debug 069 failed in check_result
control_csawmgt_debug 069 failed in run_test
control_thompson_debug 073 failed in check_result
control_thompson_debug 073 failed in run_test
control_thompson_no_aero_debug 074 failed in check_result
control_thompson_no_aero_debug 074 failed in run_test
control_thompson_extdiag_debug 075 failed in check_result
control_thompson_extdiag_debug 075 failed in run_test
control_thompson_progcld_thompson_debug 076 failed in check_result
control_thompson_progcld_thompson_debug 076 failed in run_test
regional_debug 077 failed in check_result
regional_debug 077 failed in run_test
rap_control_debug 078 failed in check_result
rap_control_debug 078 failed in run_test
rap_unified_drag_suite_debug 079 failed in check_result
rap_unified_drag_suite_debug 079 failed in run_test
rap_diag_debug 080 failed in check_result
rap_diag_debug 080 failed in run_test
rap_cires_ugwp_debug 081 failed in check_result
rap_cires_ugwp_debug 081 failed in run_test
rap_unified_ugwp_debug 082 failed in check_result
rap_unified_ugwp_debug 082 failed in run_test
rap_noah_debug 083 failed in check_result
rap_noah_debug 083 failed in run_test
rap_rrtmgp_debug 084 failed in check_result
rap_rrtmgp_debug 084 failed in run_test
rap_lndp_debug 085 failed in check_result
rap_lndp_debug 085 failed in run_test
rap_sfcdiff_debug 086 failed in check_result
rap_sfcdiff_debug 086 failed in run_test
rap_flake_debug 087 failed in check_result
rap_flake_debug 087 failed in run_test
rap_noah_sfcdiff_cires_ugwp_debug 088 failed in check_result
rap_noah_sfcdiff_cires_ugwp_debug 088 failed in run_test
rap_progcld_thompson_debug 089 failed in check_result
rap_progcld_thompson_debug 089 failed in run_test
rrfs_v1beta_debug 090 failed in check_result
rrfs_v1beta_debug 090 failed in run_test
hafs_regional_atm_thompson_gfdlsf 093 failed in check_result
hafs_regional_atm_thompson_gfdlsf 093 failed in run_test

As far as I can tell, this list corresponds to those that either use Thompson MP or Morrison-Gettelman MP. Do you have any insight as to why tests using these schemes would have different results? I.e., What if any of the changes made in fv3atm, ccpp-physics, or atmos_cubed_sphere could change the answer even when NSSL MP is not being used?

Once we figure that out, we can go ahead and add the new NSSL RT(s).

**Comment by MicroTed** [comment]
>  Other than the field_table, does NSSL need any special input data to run that we also need to stage?

I thought I had replied to this already, but must have neglected to hit the button. Only a field_table is needed, and my thinking is to have both hail and CCN prediction turned on in the namelist settings. Note that hail mass and the non-mass scalars are initialized to zero instead of 1.e30. I don't know if any extra diag_table entries would be needed, but at least to turn off anything specific to another scheme (like the aerosols in Thompson).

```
gfs_physics_nml:
    imp_physics = 17
    nssl_invertccn = .true.
    nssl_hail_on = .true.
    nssl_ccn_on = .true.
    ltaerosol = .false.

&fv_core_nml
     nwat = 7

field_table:
# added by FRE: sphum must be present in atmos
# specific humidity for moist runs
 "TRACER", "atmos_mod", "sphum"
           "longname",     "specific humidity"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
# prognostic cloud water mixing ratio
 "TRACER", "atmos_mod", "liq_wat"
           "longname",     "cloud water mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
 "TRACER", "atmos_mod", "rainwat"
           "longname",     "rain mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
 "TRACER", "atmos_mod", "ice_wat"
           "longname",     "cloud ice mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
 "TRACER", "atmos_mod", "snowwat"
           "longname",     "snow mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
 "TRACER", "atmos_mod", "graupel"
           "longname",     "graupel mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
 "TRACER", "atmos_mod", "hailwat"
           "longname",     "hail mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "water_nc"
           "longname",     "cloud droplet number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "rain_nc"
           "longname",     "rain number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "ice_nc"
           "longname",     "ice number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "snow_nc"
           "longname",     "snow number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "graupel_nc"
           "longname",     "graupel number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "hail_nc"
           "longname",     "hail number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "graupel_vol"
           "longname",     "graupel particle volume"
           "units",        "m3/kg"
       "profile_type", "fixed", "surface_value=0" /
 "TRACER", "atmos_mod", "hail_vol"
           "longname",     "hail particle volume"
           "units",        "m3/kg"
       "profile_type", "fixed", "surface_value=0" /
# ccn_nc
 "TRACER", "atmos_mod", "ccn_nc"
           "longname",     "ccn number concentration"
           "units",        "/kg"
       "profile_type", "fixed", "surface_value=0" /

# prognostic ozone mixing ratio tracer
 "TRACER", "atmos_mod", "o3mr"
           "longname",     "ozone mixing ratio"
           "units",        "kg/kg"
       "profile_type", "fixed", "surface_value=1.e30" /
# cloud droplets, water- and ice-friendly aerosols (Thompson)
# prognostic subgrid scale turbulent kinetic energy
 "TRACER", "atmos_mod", "sgs_tke"
           "longname",     "subgrid scale turbulent kinetic energy"
           "units",        "m2/s2"
       "profile_type", "fixed", "surface_value=1.e30" /

```

**Comment by MicroTed** [comment]
> What if any of the changes made in fv3atm, ccpp-physics, or atmos_cubed_sphere could change the answer even when NSSL MP is not being used?

The first thing that comes to mind is the introduction of the otsptflag array, although it is not supposed to affect results, of course. Do the non-matching results also have a convection parameterization turned on, since that is where otsptflag would come into play (in the interstitial routines). The logic "looks" OK, but I could well be missing something. 

I'll take a look at diffs and see if anything stands out.

**Comment by MicroTed** [comment]
I don't see anything else obvious that would cause difference for other physics, unless if statements within loops in atmos_cubed_sphere is affecting optimization/vectorization, e.g., fv_sg.F90, line 326:
```
             q_liq = q0(i,k,liq_wat) + q0(i,k,rainwat)
             q_sol = q0(i,k,ice_wat) + q0(i,k,snowwat) + q0(i,k,graupel)
             IF ( nwat == 7 ) THEN
             q_sol = q_sol + q0(i,k,hailwat)
             ENDIF
```
If so, a separate loop for nwat=7 could be made.

**Comment by MicroTed** [comment]
@grantfirl  I pushed an update to fv_sg.F90 to split those loops. It's a shot in the dark, but maybe see if that makes a difference for the regression tests.

**Comment by grantfirl** [comment]
@MicroTed OK, I'll try one of the failed tests to see if anything changes. I haven't had time to dig in any more, but commenting out the new otsptflag code in favor of the old logic would be something else I could try, but like you, I don't see how that would change anything. 

To answer your question from a few days ago, at least some of the failed tests use deep convection and use the new otsptflag array.

**Comment by grantfirl** [comment]
@MicroTed Your "shot in the dark" was a good one. I retried one of the failed tests and it passed with your latest change to GFDL_atmos_cubed_sphere. I'll run the rest to see if it fixes them too.

**Comment by grantfirl** [comment]
> @MicroTed Your "shot in the dark" was a good one. I retried one of the failed tests and it passed with your latest change to GFDL_atmos_cubed_sphere. I'll run the rest to see if it fixes them too.

All current UFS now passed (except for a couple of HAFS ones that had a known issue fixed in recently merged PRs). I'll work on creating new RTs today so that we can hopefully get this merged in a couple days.

**Comment by MicroTed** [comment]
> All current UFS now passed (except for a couple of HAFS ones that had a known issue fixed in recently merged PRs). I'll work on creating new RTs today so that we can hopefully get this merged in a couple days.

That great! Thanks so much for your help. If just one test is added, then I would suggest the version using hail and CCN (i.e., nwat=7, settings above).  If it seems worthwhile, a second test could run with hail turned off (nwat=6; nssl_hail_on=.false.; remove the 3 hail variables from field_table).

**Comment by grantfirl** [comment]
Please merge https://github.com/MicroTed/fv3atm/pull/3 and update this branch to the latest develop in preparation for merge.

**Comment by grantfirl** [comment]
@MicroTed Now that ccpp-physics has been updated to the latest main, you need to do the same thing here with an extra wrinkle. Locally, in your PR branch of fv3atm, add ccpp/physics and commit it to update the submodule pointer. Then, pull down latest develop, merge into your PR branch, and push to GitHub. Once this is done (and for the GFDL_atmos_cubed_sphere), I'll update the ufs-weather-model PR.

**Comment by grantfirl** [comment]
@MicroTed, I forgot to say that since GFDL_atmos_cubed_sphere is a submodule of fv3atm, you should make sure that your PR branch for that is updated to the latest main/develop (whatever they use), and do a 'git add' and 'git commit' in fv3atm to update its submodule pointer too. This should be done prior to merging in the latest develop of fv3atm, otherwise, it'll error out.

**Comment by grantfirl** [comment]
@MicroTed This PR is still showing a conflict for some reason. Any ideas?

**Comment by MicroTed** [comment]
> @MicroTed This PR is still showing a conflict for some reason. Any ideas?

I don't know why it is showing that conflict on atmos_cubed_sphere. I think my atmos_cubed_sphere is merged to the correct commit, but it's very possible I'm missing something.

**Comment by grantfirl** [comment]
> > @MicroTed This PR is still showing a conflict for some reason. Any ideas?
> 
> I don't know why it is showing that conflict on atmos_cubed_sphere. I think my atmos_cubed_sphere is merged to the correct commit, but it's very possible I'm missing something.

Please merge https://github.com/MicroTed/fv3atm/pull/4 and see if this does the trick...

**Comment by MicroTed** [comment]
Well, I thought I had already tried setting to the newer atmos_cubed_sphere commit, but there was still a conflict. Anyway, seems to be happy now. Thanks, @grantfirl !

**Comment by ChunxiZhang-NOAA** [comment]
@MicroTed, did you check if the global mean air mass (surface air pressure) was near stable with forecast lead time after adding hail category into GFDL_atmos_cubed_sphere? This check was suggested by Lucas when I implemented this scheme into IPD a few years ago.

**Comment by MicroTed** [comment]
Hi, @ChunxiZhang-NOAA, That sounds like a good idea, but I'm not actually sure off-hand how to do that. On a coarse global grid, hail won't even be produced.

**Comment by yangfanglin** [comment]
@ChunxiZhang-NOAA @MicroTed  In your forecast log file you probably can find prints like the following lines.  grep them will give the evolution of surface pressure and dry mass etc with forecast lead time.

 Total surface pressure (mb) =    985.4597
 mean dry surface pressure =    983.1575
 Total Water Vapor (kg/m**2) =   23.37231

**Comment by ChunxiZhang-NOAA** [comment]
@MicroTed  You are right. It probably needs a global run with ~ 6 or ~ 3 km grid spacing. I did a global 13 km test before, but it was likely insufficient. 
@yangfanglin Thanks Fanglin. Yes, those are the variables.

**Comment by MicroTed** [comment]
@ChunxiZhang-NOAA 
I pulled values from some regression test output (by Grant Firl).  The total sfc. pressure and total water vapor seem consistent with other suites over 24hr. For mean dry sfc. pressure, though, there is a bit of round-off change compared to constant values from comparison. It might be worth tracking down at some point why it is doing that, since the control suites seem to be constant out to many more digits.

[Ted.Mansell@hfe12 rt_8295]$ grep 'mean dry surface pressure' rrfs_v1beta/out # closest suite for comparison
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0754    
[Ted.Mansell@hfe12 rt_8295]$ grep 'mean dry surface pressure' rrfs_v1nssl/out # includes hail and CCN
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0753    
  0:  mean dry surface pressure =    983.0755    
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0755    
[Ted.Mansell@hfe12 rt_8295]$ grep 'mean dry surface pressure' rrfs_v1nssl_nohailnoccn/out
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0753    
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0754    
  0:  mean dry surface pressure =    983.0754

**Comment by ChunxiZhang-NOAA** [comment]
@MicroTed Thanks for making those comparisons. Yup, the fluctuation is small, but the control suite didn't show clear fluctuation.

File: ccpp/suites/suite_FV3_GFS_v15_nssl_mynn_lam3km.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<suite name="FV3_GFS_v15_nssl_mynn_lam3km" version="1">
```
Comment (junwang-noaa) [inline ccpp/suites/suite_FV3_GFS_v15_nssl_mynn_lam3km.xml]: Why do we add this suite file? Is there any test using this suite file?

**Review by junwang-noaa — APPROVED** [review]

**Comment by grantfirl** [comment]
@MicroTed Just a reminder that in addition to updating the submodules for ccpp/physics and gfdl_atmos_cubed_sphere with the newly merged authoritative branches (using `git add` in your fv3atm branch), please revert the changes to the .gitmodules file so that it points back to the authoritative branches. Thanks!

**Comment by junwang-noaa** [comment]
@MicroTed the CCPP physics PR and dycore PR are merged, please update your fv3atm to point to the latest official branch. Thanks. @grantfirl FYI.

**Comment by grantfirl** [comment]
@MicroTed Also consider removing the https://github.com/NOAA-EMC/fv3atm/pull/472/files#diff-2d9cb70297ca8f2d11a65437015590d50b81a3972e9931cbdf3b2d84d595d7ae and https://github.com/NOAA-EMC/fv3atm/pull/472/files#diff-d4a0a40faccabed62779161f7366eaa547180d9f655443ede4a8e1f14b2daa95 SDFs if they are not needed (e.g. if they were only for testing). If they will be used for experiments, etc., then it's fine to leave them in the repo, but if not, we don't want to clutter up the SDF directory.

**Review by grantfirl — APPROVED** [review]
@MicroTed did a thorough and careful job of adding NSSL MP. The only thing that I saw that might be good to clean up is that the new otsptflag variable was added in the interstitial DDT when it should probably be in the control DDT because it is persistent and is only ever used as input in the physics (never changed). The interstitial DDT should probably be reserved for variables that are reset every physics timestep, and otsptflag is left out of the interstitial_reset subroutines. I understand why it was put in interstitial, namely because there are many other variables, particularly related to various tracer indices, that were already there, but I'd argue that those should be in control too. Or, perhaps, have distinct interstitial_persistent and interstitial_ephemeral DDTs for variables that only are used in physics but need to stay in memory or be reset, respectively. But, this is a discussion for a later day. 

Approved, and nice job, @MicroTed!

File: ccpp/suites/suite_FV3_GFS_v15_nssl_mynn_lam3km.xml (hunk 1 above)
Comment (MicroTed) [reply ccpp/suites/suite_FV3_GFS_v15_nssl_mynn_lam3km.xml]: It's one that we have been using, but I suppose it doesn't need to be in this repository. I can take it out.

Comment (grantfirl) [reply ccpp/suites/suite_FV3_GFS_v15_nssl_mynn_lam3km.xml]: If you're using it, I think it is important to keep it in -- probably more important than the one used in RTs.

Comment (junwang-noaa) [reply ccpp/suites/suite_FV3_GFS_v15_nssl_mynn_lam3km.xml]: Yes, it's good to keep it if you are using it. Thanks

**Comment by MicroTed** [comment]
@grantfirl  I think I updated per your suggestions. Let me know if I missed anything.

As for the otsptflag thing, I'm open to anything there.

**Comment by grantfirl** [comment]
> 

Yep, everything looks good. I think the organization of which DDTs various variables belong in requires a larger conversation and is only tangentially related to anything that you've done here.

**Pull request merged**